clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/InactiveTopicPartitionChecker.java

```diff
@@ -0,0 +1,221 @@
+      executorService.shutdown();
+      try {
+        if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
+          executorService.shutdownNow();
+        }
+      } catch (InterruptedException e) {
+        executorService.shutdownNow();
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  /**
+   * Checks for inactive topic partitions and handles them appropriately.
+   *
+   * <p>This method is called periodically by the scheduled executor service and performs
+   * the following operations for each consumer:
+   *
+   * <ol>
+   *   <li><b>Detection:</b> Identifies currently inactive topic partitions by comparing
+   *       the per-topic-partition last successful poll timestamp against the configured threshold.
+   *       Skips topic partitions that are already paused to avoid redundant checks.</li>
+   *   <li><b>Resume:</b> Resumes previously paused topic partitions that are no longer inactive</li>
+   *   <li><b>Pause:</b> Pauses newly detected inactive topic partitions and adds them to the tracking set</li>
+   * </ol>
+   *
+   * <p>A topic partition is considered inactive if:
+   * <ul>
+   *   <li>The time since its last successful poll exceeds {@code inactiveTopicPartitionThresholdInMs}</li>
+   *   <li>It has never been successfully polled (lastSuccessfulPollTimestamp == -1)</li>
+   * </ul>
+   *
+   * <p>The method includes comprehensive error handling to ensure that exceptions
+   * don't break the periodic execution schedule.
+   *
+   * <p><b>Note:</b> This implementation uses per-topic-partition statistics from {@link ConsumptionTask#getPartitionStats(PubSubTopicPartition)}
+   * to get accurate activity information for each topic partition individually, rather than using a global
+   * last poll timestamp for the entire consumer.
+   */
+  void checkInactiveTopicPartition() {
+    try {
+      long currentTime = System.currentTimeMillis();
+
+      // Process each consumer: check for inactive topic-partitions, then pause/resume as needed
+      for (Map.Entry<SharedKafkaConsumer, ConsumptionTask> entry: getConsumerToConsumptionTask().entrySet()) {
+        SharedKafkaConsumer consumer = entry.getKey();
+        ConsumptionTask task = entry.getValue();
+        Set<PubSubTopicPartition> previouslyPausedTopicPartitions =
+            getConsumerToPausedTopicPartitionsMap().computeIfAbsent(consumer, x -> new HashSet<>());
+        Set<PubSubTopicPartition> assignedTopicPartitions = consumer.getAssignment();
+        Set<PubSubTopicPartition> currentlyInactiveTopicPartitions = new HashSet<>();
+
+        LOGGER.info(
+            "Checking inactive topic partition for consumer task: {} with current assignments: {}",
+            task.getTaskIdStr(),
+            assignedTopicPartitions);
+        // Step 1: Check all assigned topic-partitions for inactivity and collect currently inactive ones
+        for (PubSubTopicPartition topicPartition: assignedTopicPartitions) {
```

Comment:
Instead of maintaining separate loops, we could collapse this into a single loop. That way, all actions for a given topic partition are grouped together, which makes the flow easier to follow and avoids multiple iterations over these collections.

For each partition in the loop:
	1.	Check whether it is active or inactive.
	2.	Pause if it is inactive.
	3.	Resume if the elapsed time condition is met.
	4.	Move on to the next partition.

WDYT?

Reply:
As chatted offline, I actually want to further reduce log rate, so I will instead log the set of the paused/resumed TPs in one message instead, so I will need to keep the current loop structure which is also generated by AI :)